Remove StreamBitmap-related dead code

armenatzoglou opened the pull request:
If GPDB evaluates a query that needs to access an index, e.g., btree, then at least one `Bitmap Index Scan` will appear in query's plan. `MultiExecBitmapIndexScan` function will be invoked to construct the bitmap. If a query contains a `WHERE` clause similar to `d in (1,2)`, where a bitmap index has been created on column `d`), then instead of creating a simple bitmap, `MultiExecBitmapIndexScan` will generate a composite bitmap that ORs all the bitmaps that satisfy the query condition. In our example, `MultiExecBitmapIndexScan` (in particular `bmgetmulti`) will OR the bitmaps of `d = 1` and `d = 2`. 

To OR two bitmaps `A` and `B`, `A` should be a `StreamBitmap` (and not a `HashBitmap`). 

In this PR, we remove code that was executed when `MultiExecBitmapIndexScan` has to generate a composite bitmap of `A` and `B`, where `A` is a `StreamBitmap` and `B` is a `HashBitmap` generated when accessing either a btree or a gin or a gist or a hash index. It seems that this is not a possible scenario because i) `A` can be a `StreamBitmap` only if a bitmap index has been constructed on a particular column, and ii) to the best of our knowledge we cannot have a single `Bitmap Index Scan` that will access both a Bitmap index and an index of another type e.g., btree. 

We cannot find a query that exercises the deleted code. Also, the code is not exercised in ICG, which is green after the deletion of the code. Please let us know if you are aware of a scenario where this code is exercised or if you agree that is dead code.

 @foyzur @ashwinstar @asimrp @d @hlinnaka

hlinnaka commented:
Agreed. An index AM's getmulti function can assume that the bitmap that it's given as argument is of the same type as what the function constructs itself. bmgetmulti() already makes that assumption, and throws an error if it's passed a HashBitmap (it constructs StreamBitmaps itself).

We should add comments to mention this though. Not sure where. In the upstream, this would belong in the documentation of the index AM API, in doc/src/sgml/indexam.sgml, but we don't have that in GPDB. Perhaps a largish comment in index_getmulti() would be in order, to explain what the function does. The other indexam API functions in GPDB are identical to the upstream, but getmulti() is not, so it would be good to document that.

hlinnaka commented on src/backend/access/gin/ginget.c:
```diff
@@ -485,13 +485,6 @@ gingetmulti(PG_FUNCTION_ARGS)
  		else
  			break;
 	}	
- 
```
Also no need for the IsA(n, StreamBitmap) check earlier in this function. Although an Assert or "if (!IsA(n, StreamBitmap) elog(ERROR, "...")" would probably be  in order. Same with all the other index types.

armenatzoglou commented:
@hlinnaka Good points. We added elog(ERROR) for every index so that we are consistent. Also, we added the comments that you suggested in index_getmulti. Thanks

armenatzoglou commented:
Pushed to master 77ea2a503ac8f26ab7e436ff991323802902107b. Thanks a lot for the review.
